modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/pagemem/ExponentialBackoff.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2021 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.persistence.pagemem;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Implements exponential backoff logic. Contains a counter and on each {@link #nextDuration()} increments it.
+ * May be reset using {@link #reset()}.
+ */
+class ExponentialBackoff {
+    /**
+     * Starting backoff duration.
+     */
+    private final long startingBackoffNanos;
+
+    /**
+     * Backoff ratio. Each next duration will be this times longer.
+     */
+    private final double backoffRatio;
+
+    /**
+     * Exponential backoff counter.
+     */
+    private final AtomicInteger exponentialBackoffCounter = new AtomicInteger(0);
```

Comment:
Could you please elaborate? I don't seem to understand what change is suggested or what is wrong with the current code at this line.

Reply:
I think that the current code is absolutely fine. I'm suggesting a minor "improvement": we currently store the exponent, which we then use to compute the backoff value using the `pow` function. Instead we can cache the previous backoff value and simply multiply it by `backoffRatio` on each call. This is just an idea, not a requirement

Reply:
Oh, I see what you mean. Yes, we could try storing the current backoff value in an `AtomicLong` and update it using `getAndUpdate(prev -> (long) (prev * backoffRatio))`. But this would compute slightly different backoff values (because we would accumulate rounding error on each step), which does not see important, but still different. Also, under contention, the new code would have to do some wasted multiplications, while now it only wastes increments (which are cheaper). So it is not that obvious whether it would perform better or not under contention.

Let's leave it as is for now and improve it later when we have time for this specific improvement (including perf tests).